Release blog post 1.14 draft iteration (#468)
apache/flink-web: _posts/2021-09-21-release-1.14.0.md

---
layout: post 
title:  "Apache Flink 1.14.0 Release Announcement"
date: 2021-09-21T08:00:00.000Z 
categories: news 
authors:
- joemoe:
  name: "Johannes Moser"

excerpt: The Apache Flink community is excited to announce the release of Flink 1.14.0! Around xxx contributors worked on over xxxx issues to TODO.
---

> Just a couple of days ago the Apache Software Foundation announced its annual report and Apache
> Flink was again in the Top 5 of the most active projects in all relevant categories. This remarkable
> activity is also reflected in this new 1.14.0 release. Once again, more than 200 contributors worked on
> over 1,000 issues. We are proud of how this community is consistently moving the project forward.

Reviewer: ```suggestion
The Apache Software Foundation recently released its annual report and Apache Flink once again made it on the list of the top 5 most active projects! We think that this recognition is reflected in the new 1.14.0 release. More than 200 contributors worked on over 1,000 issues. We are proud of how this community is consistently moving the project forward.
```

Author: Merged this with slight adjustments.

The release brings many cool improvements, from SQL to connectors, checkpointing, and PyFlink.
A big area of changes in this release is the integrated streaming & batch experience. We believe

> that unbounded stream processing goes hand-in-hand with bounded- and batch processing tasks in practice.

Reviewer 2: ```suggestion
that unbounded stream processing goes hand-in-hand with bounded and batch processing tasks in practice.
```

Author: I think it is correct as it is. The dash indicates that there is another work coming (here "processing" that logically also belongs to this word).

Long form: "... bounded processing and batch processing"
Short form: "... bounded- and batch processing"

> The release brings many cool improvements, from SQL to connectors, checkpointing, and PyFlink.
> A big area of changes in this release is the integrated streaming & batch experience. We believe
> that unbounded stream processing goes hand-in-hand with bounded- and batch processing tasks in practice.
> Data exploration when developing new applications, bootstrapping state for new applications, training
> models to be applied in the streaming application, re-processing data after fixes/upgrades, and many
> other use cases require processing historic data from various sources next to the streaming data.

Reviewer: ```suggestion
This release brings many new features and improvements in areas such as the SQL API, more connector support, checkpointing, and PyFlink. But most notable are the changes to the integrated streaming and batch experience. We believe that, in practice, unbounded stream processing should work seamlessly with bounded processing and batch processing tasks. Many use cases require processing historic data from various sources alongside streaming data (i.e. data exploration when developing new applications, bootstrapping state for new applications, training models to be applied in the streaming application, re-processing data after fixes/upgrades).  
```

Author: Merged this with slight adjustments.

> In Flink 1.14, we finally made it possible to **mix bounded and unbounded streams in an application**;
> Flink can now take checkpoints of applications that is partially running and partially finished (some
> operators reached the end of the bounded inputs). Additionally, **bounded streams now take a final checkpoint**
> when reaching their end to ensure smooth committing of results in sinks.
> The **batch execution mode now works for programs that mix DataStream & Table/SQL** (previously only
> pure Table/SQL or DataStream programs). The unified Source and Sink APIs have made strides ahead,
> we started **consolidating the connector ecosystem around the unified APIs**, and added a **hybrid source**
> that can bridge between multiple storage systems, like start reading old data from S3 and switch over
> to Kafka later.
> 
> Furthermore, this release takes another step in our initiative of making Flink more self-tuning and
> to require less Stream-Processor-specific knowledge to operate. We started that initiative in the previous
> release with [Reactive Scaling](/news/2021/05/03/release-1.13.0.html#reactive-scaling) and are now
> adding **automatic network memory tuning** (*a.k.a. Buffer Debloating*). This feature speeds up checkpoints
> under load without sacrificing performance or increasing checkpoint size, by continuously adjusting the
> network buffering to ensure best throughput while having minimal in-flight data. See the
> [Buffer Debloating section](#buffer-debloating) for details.

Reviewer: ```suggestion
In Flink 1.14, it is now possible to **mix bounded and unbounded streams in an application**. 
To achieve this, Flink now takes checkpoints of applications that are partially running and partially finished (when some operators reach the end of the bounded inputs). Additionally, **bounded streams now take a final checkpoint** when reaching their end to ensure smooth result commits to sinks.

The **batch execution mode now works for programs that use a mixture of the DataStream API and the Table & SQL APIs**. 

The unified Data Source and Data Sink APIs have also been greatly improved. We started **consolidating the connector ecosystem around the unified APIs** and added a **hybrid source** that can bridge multiple storage systems.  You can now do things like read old data from Amazon S3 and then switch over to Apache Kafka.

In addition, this release furthers our initiative in making Flink more self-tuning and user-friendly. We aim to bring Flink to a wider audience and make it easier to operate without necessarily requiring a lot of knowledge in stream processing. We started this initiative in the previous release with [reactive scaling](/news/2021/05/03/release-1.13.0.html#reactive-scaling) and are now adding **automatic network memory tuning** (*a.k.a. Buffer Debloating*). This feature speeds up checkpoints under high load without sacrificing performance or increasing checkpoint size by continuously adjusting network buffers to ensure the best throughput while having minimal in-flight data. See the [Buffer Debloating section](#buffer-debloating) for more details.
```

Author: Merged this with slight adjustments.

There are many more improvements and new additions throughout various components, as we discuss below.
We also had to say goodbye to some features that have been superceded by newer ones in recent releases,

> most prominently we are **removing the old SQL execution engine**.

Reviewer 2: Perhaps dropping Mesos supports is another thing to mention?

Author: Good catch, adding this.

> There are many more improvements and new additions throughout various components, as we discuss below.
> We also had to say goodbye to some features that have been superceded by newer ones in recent releases,
> most prominently we are **removing the old SQL execution engine**.
> 
> We hope you like the new release and we'd be eager to learn about your experience with it, which yet
> unsolved problems it solves, what new use-cases it unlocks for you.

Reviewer: ```suggestion
We had to say goodbye to some features that have been superseded by newer ones in recent releases, such as the **removal of the old SQL execution engine** and **dropping support for Apache Mesos**.

There are a lot more improvements and new additions throughout various components in Flink and they will be discussed in more detail below. We hope that you like the new release and that it helps you solve new problems and unlocks new use cases. :)  And as always, we would appreciate any sort of feedback. 
```

{% toc %}

> # Unified Batch and Stream Processing experience
> 
> One of Flink's unique characteristics is how it integrates streaming and batch processing,
> using common unified APIs, and a runtime that supports multiple execution paradigms.
> 
> As motivated in the introduction, we believe Streaming and Batch always go hand in hand. This quote from
> a [report on Facebook's streaming infrastructure](https://research.fb.com/wp-content/uploads/2016/11/realtime_data_processing_at_facebook.pdf)
> echos this sentiment nicely.

Reviewer: ```suggestion
# The Unified Batch and Stream Processing Experience

One of Flink's unique characteristics is how it integrates batch and stream 
processing by using unified APIs and a runtime that supports multiple execution paradigms.

We believe that batch and stream processing are not separate paradigms and that batch
is a special case of streaming. This quote from
a [report on Facebook's streaming infrastructure](https://research.fb.com/wp-content/uploads/2016/11/realtime_data_processing_at_facebook.pdf)
echos this sentiment nicely.
```

> Streaming versus batch processing is not an either/or decision. Originally, all data warehouse
> processing at Facebook was batch processing. We began developing Puma and Swift about five years
> ago. As we showed in Section [...], using a mix of streaming and batch processing can speed up
> long pipelines by hours.

Having both the real-time and the historic computations in the same engine also ensures consistency
between semantics and makes results well comparable. Here is an [article by Alibaba](https://www.ververica.com/blog/apache-flinks-stream-batch-unification-powers-alibabas-11.11-in-2020)
about unifying business reporting with Apache Flink and getting consistent reports that way.

> While unified streaming & batch are already possible in earlier versions, this release bring a

Reviewer 2: ```suggestion
While unified streaming & batch are already possible in earlier versions, this release brings
```

Author: fixed

> Having both the real-time and the historic computations in the same engine also ensures consistency
> between semantics and makes results well comparable. Here is an [article by Alibaba](https://www.ververica.com/blog/apache-flinks-stream-batch-unification-powers-alibabas-11.11-in-2020)
> about unifying business reporting with Apache Flink and getting consistent reports that way.
> 
> While unified streaming & batch are already possible in earlier versions, this release bring a
> some features that unlock new use cases, as well as a series of quality-of-life improvements.

Reviewer: ```suggestion
Keeping both real-time and the historic computations in the same engine also ensures consistency
between semantics and makes results more accurate and meaningful. Here is an [article by Alibaba](https://www.ververica.com/blog/apache-flinks-stream-batch-unification-powers-alibabas-11.11-in-2020)
about obtaining consistent reports through unifying business reporting with Apache Flink.

Although unified batch and stream processing was already possible in earlier versions of Flink, this release introduces features that unlock new use cases, as well as improvements in user experience.
```

## Checkpointing and Bounded Streams

> Flink's checkpointing mechanism could originally only draw checkpoints when all tasks in an application's
> DAG were running. That means that mixed bounded/unbounded applications were not really possible.
> In addition, applications on bounded inputs that were executed in a streaming way (not in a batch way)
> stopped checkpointing towards the end, when some tasks finished. Lingering data for exactly-once
> sinks was the result, because in the absence of checkpoints, the latest output data was not committed.
> 
> With [FLIP-147](https://cwiki.apache.org/confluence/display/FLINK/FLIP-147%3A+Support+Checkpoints+After+Tasks+Finished)
> Flink now supports checkpoints after tasks are finished, and takes a final checkpoint at the end of a
> bounded stream, ensuring that all sink data is committed before the job ends (similar to how
> *stop-with-savepoint* behaves).
> 
> To activate this feature, add `execution.checkpointing.checkpoints-after-tasks-finish.enabled: true`
> to your configuration. Keeping with the tradition of making big new features opt-in for the first release,
> the feature is not active by default in Flink 1.14. We expect it to become the default in the next release.
> 
> As a side note: Why would we even want to execute applications over bounded streams in a streaming way,
> rather than in a batch-y way? There can be many reasons, like (a) the sink that you write to needs to be
> written to in a streaming way (like the Kafka Sink), or (b) that you do want to exploit the
> streaming-inherent quasi-ordering-by-time, such as in the [Kappa+ Architecture](https://youtu.be/4qSlsYogALo?t=666).

Reviewer: ```suggestion
Flink's checkpointing mechanism could previously only create checkpoints when all tasks in an application's
directed acyclic graph (DAG) were running. This meant that applications using both bounded and unbounded data processing were not really possible. In addition, applications with bounded inputs that were executed in an unbounded manner stopped checkpointing towards the end of processing, which can result in only some tasks finishing. Without checkpoints, the latest outputs were not committed, resulting in lingering data for exactly-once
sinks.

With [FLIP-147](https://cwiki.apache.org/confluence/display/FLINK/FLIP-147%3A+Support+Checkpoints+After+Tasks+Finished),
Flink now supports checkpoints after tasks are finished and takes a final checkpoint at the end of a
bounded stream, ensuring that all output data is committed before the job ends (similar to how
*stop-with-savepoint* behaves).

To activate this feature, add `execution.checkpointing.checkpoints-after-tasks-finish.enabled: true`
to your configuration. Keeping with the opt-in tradition for big and new features,
this is not activated by default in Flink 1.14. We expect it to become the default mode in the next release.

Note: You may want to execute applications over bounded streams in an unbounded way for many reasons. The sink being used may need to be written to in a streaming manner (i.e. Kafka sink) or you may not want to exploit the streaming-inherent quasi-ordering-by-time, such as in the [Kappa+ Architecture](https://youtu.be/4qSlsYogALo?t=666).
```

> ## Mixing Batch DataStream API and Table API/SQL
> 
> SQL and the Table API are becoming the default starting points for new projects. The declarative
> nature and richness of built-in types and operations make it easy to develop applications fast.
> It is not uncommon, though, for developers to eventually hit the limit of SQL's expressiveness for
> certain types of event-driven business logic (or hit the point when it becomes grotesque to express
> that logic in SQL).
> 
> At that point, the natural step is to blend in a piece of stateful DataStream API logic, before
> switching back to SQL again.
> 
> In Flink 1.14, bounded batch-executed SQL/Table programs can convert their intermediate
> Tables to a DataStream, apply some DataSteam API operations, and convert it back to a Table. Flink builds
> a dataflow DAG mixing declarative optimized SQL execution with batch-executed DataStream logic.
> Check out the [relevant docs](https://nightlies.apache.org/flink/flink-docs-release-1.14/docs/dev/table/data_stream_api/#converting-between-datastream-and-table) for details.

Reviewer: ```suggestion
## Mixing the DataStream API and the SQL/Table API

SQL and the Table API are becoming the default starting points for new projects. The declarative
nature and richness of built-in types and operations make it easy to develop applications fast.
It is not uncommon, however, for developers to eventually hit the limit of SQL's expressiveness for
certain types of event-driven business logic (or hit the point when it becomes tricky to express
that logic in SQL). At that point, the natural step is to blend in some stateful DataStream API logic before
switching back to SQL again.

In Flink 1.14, bounded batch-executed SQL/Table API programs can convert their intermediate
Tables to a DataStream, apply some DataStream API operations, and convert it back to a Table. Under the hood, Flink builds a dataflow DAG that mixes declarative optimized SQL execution with batch-executed DataStream logic.
Check out the [documentation](https://nightlies.apache.org/flink/flink-docs-release-1.14/docs/dev/table/data_stream_api/#converting-between-datastream-and-table) for more details.
```

> ## Hybrid Source
> 
> The new [Hybrid Source](https://nightlies.apache.org/flink/flink-docs-release-1.14/docs/connectors/datastream/hybridsource/)
> produces a combined stream from other sources, by reading those other sources one after the other,
> seamlessly switching over from one to the other.
> 
> With the Hybrid Source, you can read for example from tiered storage setups as if there was one
> stream that spans everything. Consider the example below: New data lands in Kafa and is eventually
> migrated to S3 (typically in compressed columnar format, for cost efficiency and performance).
> The Hybrid Source can read a stream that starts all the way in the past on S3 and transitions over
> to the real-time data in Kafka.
> 
> <figure style="align-content: center">
>   <img src="{{ site.baseurl }}/img/blog/2021-09-25-release-1.14.0/hybrid_source.png" style="display: block; margin-left: auto; margin-right: auto; width: 600px"/>
> </figure>
> 
> We believe that this is an exciting step in realizing the full promise of logs and the *Kappa Architecture.*
> Even if older parts of a log of events are physically migrated to different storage
> (because cheaper, better compression, faster to read) you can still treat and process it as one
> contiguous log.
> 
> The Hybrid Source foundation is in Flink 1.14. Over the next releases, we expect to add more
> utilities and patterns for typical switching strategies.

Reviewer: ```suggestion
## Hybrid Data Sources

The new [Hybrid Source](https://nightlies.apache.org/flink/flink-docs-release-1.14/docs/connectors/datastream/hybridsource/) feature
produces a combined stream from multiple sources. It reads data sources sequentially and
seamlessly switches from one source to the other.

With Hybrid Source, you can read data from tiered storage setups as if there was one
stream. This can be useful for many scenarios.  For example, new data may land in a Kafka stream and is eventually
migrated to an S3 bucket (typically in compressed columnar format for cost efficiency and performance).
The Hybrid Source can read a stream that began on a S3 source and then subsequently transitioned over
to a real-time data stream in Kafka.

<figure style="align-content: center">
  <img src="{{ site.baseurl }}/img/blog/2021-09-25-release-1.14.0/hybrid_source.png" style="display: block; margin-left: auto; margin-right: auto; width: 600px"/>
</figure>

We believe that this is an exciting step in realizing the full potential of logs and the *Kappa Architecture.*
Even if older parts of an event log are physically migrated to different storage
(for reasons such as cost, better compression, faster reads) you can still treat and process it as one
contiguous log.

The foundation for hybrid data sources has been laid in Flink 1.14. Over the next releases, we expect to add more
utilities and patterns for typical switching strategies.
```

> ## Consolidating Sources and Sink
> 
> With the new unified (streaming/batch) source and sink APIs becoming stable, we started the
> big effort to consolidate all connectors around those APIs. At the same time, are
> better aligning connectors between DataStream and SQL/Table API. First are the *Kafka* and
> *File* Soures and Sinks for the DataStream API.
> 
> The result of this effort (that we expect to span at least 1-2 futher releases) will be a much
> more smooth and consistent experience for Flink users when connecting to external systems;
> something were we need to acknowledge that there is room for improvement in Flink.

Reviewer: ```suggestion
## Consolidating Sources and Sinks

With the new unified source and sink APIs becoming more stable, we 
began the effort of consolidating all connectors around these APIs. There are also
better connector alignments between the DataStream API and the SQL/Table API, with visible changes in the *Kafka* and 
*File* sources and sinks for the DataStream API.

The result of this effort (which we expect to span at least 1-2 further releases) will be a much
smoother and more consistent experience for Flink users when connecting to external systems.
```

# Improvements to Operations

## Buffer debloating

*Buffer Debloating* is a new technology in Flink that automatically tunes the usage
of network memory to ensure throughput, while minimizing in-flight data and thus minimizing
checkpoint latency and cost.

Apache Flink buffers a certain amount of data in its network stack to be able to utilize the

> bandwidth of fast networks. A Flink application running under high trhoughput uses some of

Reviewer 3: ```suggestion
bandwidth of fast networks. A Flink application running under high throughput uses some of
```

<figure style="align-content: center">
  <img src="{{ site.baseurl }}/img/blog/2021-09-25-release-1.14.0/buffer_debloating.png" style="display: block; margin-left: auto; margin-right: auto; width: 600px"/>
</figure>

Buffer Deloating acts as a complementary feature, or even alternative, to unaligned checkpoints.
Checkout [the docs](https://nightlies.apache.org/flink/flink-docs-release-1.14/docs/deployment/memory/network_mem_tuning/#the-buffer-debloating-mechanism)
to see how to activate this feature.

## Fine-grained Resource Management

The new *Fine-grained Resource Management* is an advanced feature to increase the resource
utilization of large shared clusters.

Flink clusters execute various data processing workloads, and different data processing steps typically need
different resources: compute resource, memory, etc. For example a filter in SQL needs different
resources than a join. By default, Flink manages resources in coarse-grained
units called 'slots', which are slices of a TaskManager's resources. A slot is then filled with tasks
that use those resources. Through *'slot sharing groups'*, users can influence how the tasks are put
into slots.

> With fine-grained resource management, TaskManager slots now dynamically sized. Transformations and Operators

Reviewer 3: ```suggestion
With fine-grained resource management, TaskManager slots are now dynamically sized. Transformations and Operators
```

orchestration layer within Flink. Hmmmm, why would we even do that, is this not better handled directly by
Kubernetes or Yarn? For most cases, that is in fact true. But we have seen that in large shared clusters,
resource requirements change so fast (especially in SQL jobs where each operator needs different resources),
that a lot of efficiency gets lost waiting for Yarn and K8s to fullfill the resource requests (taking seconds,
where Flink internally does it in milliseconds).

This feature is mainly useful for large platform deployments where Flink runs many batch
jobs of varying kind and manages the resources actively (dynamically requesting and releasing resources
from Kubernetes or Yarn, as opposted to opaque reactive deployments). Please refer to the
[Fine-grained Resource Management Docs](https://nightlies.apache.org/flink/flink-docs-release-1.14/docs/deployment/finegrained_resource/)
for details on how to use this feature.

<figure style="align-content: center">
  <img src="{{ site.baseurl }}/img/blog/2021-09-25-release-1.14.0/fine_grained_resource_management.png" style="display: block; margin-left: auto; margin-right: auto; width: 600px"/>
</figure>

> Alibaba's internal SQL platform on top of Flink and has used thich mechanism for some years now and

Reviewer 3: ```suggestion
Alibaba's internal SQL platform on top of Flink and has used this mechanism for some years now and
```

has seen significantly increase the resource utilization of the cluster. This feature was open sourced
to help other users that build large platforms on top of Flink to optimize resource footprints.

> 

Reviewer 2: 1. This feature serves both long running streaming jobs and short lived batch / bounded-stream jobs. IMHO, it's is more helpful to the long running jobs than to the batch jobs. Short lived jobs tend to prioritize scheduling overhead over resource utilization.

2. Why not to leverage the orchestration of K8s / Yarn directory?
- Avoid waiting for pod / container allocation and initialization (as mentioned) is one of the reasons.
- Slots are sometimes small. Leveraging multi-slot TMs is more efficient than requesting many TMs, by sharing the framework overhead. Moreover, having too many small pods / containers also create significant tensions for K8s / Yarn.
- The key of fine-grained resource management is to have dynamic sized slots. Slicing these slots from TMs with the same predefined size is only the first step, with minimum changes to the previous mechanism. A potential next step would be having dynamic sized TMs, where the sizes are derived from the requested slots. In other words, having dynamic sized slots / TMs is a precondition for better leveraging orchestration of K8s / Yarn.

3. I'd suggest not to emphasize the SQL use cases. This feature is only available to DataStream jobs in Flink 1.14.

Reviewer 2: This figure might help demonstrate the benefit of this feature, comparing with coarse-grained resource management.
https://drive.google.com/file/d/1-xpwluhzNy_N44a5Z0OOygIduQgkfWR9/view?usp=sharing

Author: I adjusted the text and used your new figure (removed the non-slot-sharing part, to create less confusion).

# Connectors

## Connector metrics

Flink 1.14 standardized the metrics for connectors ([FLIP-33](https://cwiki.apache.org/confluence/display/FLINK/FLIP-33%3A+Standardize+Connector+Metrics))
The community will gradually pull them through all connectors, as we rework them
onto the new unified APIs over the next releases. In Flink 1.14, we cover the Kafka connector,
and partially the File connectors.

Connectors are the entry and the exit point for data in Flink job. If a job is not running as
expected, the connector telemetry is among first parts to check. We believe this will become
a nice improvement for everyone operating production Flink applications.

> ## Pulsar Source

Reviewer 2: Maybe `Pulsar Connector`, it there're both source and sink.

Author: I updated this text. Thanks for the comment!

Flink added a [new Source and Sink](https://nightlies.apache.org/flink/flink-docs-release-1.14/docs/connectors/datastream/pulsar/).
for [Apache Pulsar](https://pulsar.apache.org/). Pulsar is a distributed messaging and streaming
system.

The connector currently supports the DataStream API. Table API/SQL bindings are expected to be

> contributed next.

Reviewer 4: I've rewritten this section to explain in a little detail the features of Pulsar. I hope it could be accepted.

```
In this release, Flink added the [Apache Pulsar](https://pulsar.apache.org/) connector. The Pulsar connector reads data from Pulsar topics in both the streaming and the batch running modes. With the support of the transaction functionality (introduced in Pulsar 2.8.0), the Pulsar connector provides exactly-once delivery semantic to ensure that a message is delivered exactly once to the consumer, even if a producer retries sending the message. 

And, to ensure message ordering and scaling and serve different purposes, the Pulsar connector also supports four subscription types:

* [Exclusive](https://pulsar.apache.org/docs/en/concepts-messaging/#exclusive)
* [Shared](https://pulsar.apache.org/docs/en/concepts-messaging/#shared)
* [Failover](https://pulsar.apache.org/docs/en/concepts-messaging/#failover)
* [Key-Shared](https://pulsar.apache.org/docs/en/concepts-messaging/#key_shared)

In the future, we will fully implement the Pulsar Sink and the Table API/SQL bindings.

For details about how to use the Pulsar connector, see [Apache Pulsar Connector](https://nightlies.apache.org/flink/flink-docs-release-1.14/docs/connectors/datastream/pulsar/#apache-pulsar-connector).
```

Author: Thanks, I am adding this to the blog post.

# Goodbye Legacy SQL Engine

Maintaining an open source project also means saying good-bye to some beloved features eventually.
When we added the Blink SQL Engine to Flink more than two years ago, it was clear that it would
eventually replace the previous SQL engine. Blink was faster and more feature complete.

For a year now, Blink has been the default SQL engine. With Flink 1.14 we finally remove all
code from the previous SQL engine. Not only does this help contributors to be faster with their
development, but it means we can also drop a plethora of outdated interfaces and reduce confusion
for users which interfaces to use when implementing custom connectors or functions.

# Python

The python API also got more observability features and optimisations.

> TBD.

Reviewer 5: ```suggestion
Python function chaining is a technology which combines multiple Python 
functions into one instance during execution. It could avoid unnecessary 
RPC communications between the Java and Python processes, eliminate 
unnecessary serialization/deserialization overhead and improve the overall 
performance. Python function chaining is already available for Python UDFs 
defined in the Python Table API & SQL. In Flink 1.14, it's also available for the 
Python functions defined in Python DataStream API.

Python functions are executed in a separate Python process in PyFlink. This 
makes it difficult to debug for users although they could use Python's builtin 
logging module in their Python functions. In Flink 1.14, we have addressed 
this problem by introducing the loopback mode which is activated by default in
local deployment. Under loopback mode, user-defined Python functions will be
executed in the Python process of the client side instead of launching new 
Python processes. Users could debug their Python functions easily just by setting
breakpoints in the IDEs before executing the job.

There are also many other improvements, e.g. it has supported to execute PyFlink 
jobs in YARN application mode, supported tgz files as Python archives, allowed to 
configure the Python interpreter of client side, etc. Check out the [relevant docs](https://nightlies.apache.org/flink/flink-docs-release-1.14/docs/dev/python/overview/) 
for details.
```

Author: I incorporated (and adjusted) this text.

Flink now supports taking checkpoints of applications that are partially running and partially finished (some
operators reached the end of the bounded inputs). Additionally, **bounded streams now take a final checkpoint**
when reaching their end to ensure smooth committing of results in sinks.

The **batch execution mode now supports programs that use a mixture of the DataStream API and the SQL/Table API**
(previously only pure Table/SQL or DataStream programs).

The unified Source and Sink APIs have gotten an update, and we started **consolidating the connector ecosystem around the unified APIs**. We added a new **hybrid source** can bridge between multiple storage systems.
You can now do things like read old data from Amazon S3 and then switch over to Apache Kafka.

Furthermore, this release takes another step in our initiative of making Flink more self-tuning and
to require less Stream-Processor-specific knowledge to operate. We started that initiative in the previous
release with [Reactive Scaling](/news/2021/05/03/release-1.13.0.html#reactive-scaling) and are now
adding **automatic network memory tuning** (*a.k.a. Buffer Debloating*). This feature speeds up checkpoints

> under load without sacrificing performance or increasing checkpoint size, by continuously adjusting the

Reviewer 6: > without sacrificing performance

That's unfortunately not true. It's causing a couple of % slow down, that might be hard to get rid of. What's worse, is that it looks like this performance drop is visible even in heavily back pressured jobs, with low throughputs and long time to process single record. In other words, it doesn't slow down max theoretical throughput, but potentially it might slow down every job.

Author: I'll make a small adjustment.

network buffering to ensure best throughput while having minimal in-flight data. See the
[Buffer Debloating section](#buffer-debloating) for details.

In addition, this release furthers our initiative in making Flink more self-tuning. We aim make Flink
easier to operate without necessarily requiring a lot of Stream-Processor-specific knowledge.
We started this initiative in the previous release with [reactive scaling](/news/2021/05/03/release-1.13.0.html#reactive-scaling)
and are now adding **automatic network memory tuning** (*a.k.a. Buffer Debloating*).
This feature speeds up checkpoints under high load without sacrificing performance or
increasing checkpoint size by continuously adjusting network buffers to ensure the best
throughput while having minimal in-flight data. See the [Buffer Debloating section](#buffer-debloating)

> for more details.

Reviewer 6: Those paragraphs are duplicated

Author: fixed

all of that memory. Aligned checkpoints flow with the data through the network buffers in milliseconds.

When a Flink application becomes (temporarily) backpressured (for example when being backpressured
by an external system, or when hitting skewed records), there is typically now a lot more data in
the network buffers than is necessary for the network to support the application's current throughput
(which is lowered due to backpressure). There is even an adverse effect: more buffered data means
the checkpoints need to do more work. Aligned checkpoint barriers need to wait for more data to be
processed, unaligned checkpoints need to persist more in-flight data.

This observation that under backpressure there is more data in the network buffers than than necessary is
where *Buffer Debloating* starts: It changes the network stack from keeping up to X bytes of data
to keeping data that is worth X milliseconds of receiver computing time. With the default setting
of 1000 milliseconds, that means the network stack will buffer as much data as the receiving task can
process in 1000 milliseconds. This value is constantly measured and adjusted, so the system keeps

> this characteristic under even under varying conditions. As a result, Flink can now provide

Reviewer 6: ```suggestion
this characteristic even under varying conditions. As a result, Flink can now provide
```

Author: Fixed, thanks

stable and predictable alignment times for aligned checkpoints under backpressure, and can vastly
reduce the amount of in-flight data stored in unaliged checkpoints under backpressure.

<figure style="align-content: center">
  <img src="{{ site.baseurl }}/img/blog/2021-09-25-release-1.14.0/buffer_debloating.png" style="display: block; margin-left: auto; margin-right: auto; width: 600px"/>
</figure>

Buffer Deloating acts as a complementary feature, or even alternative, to unaligned checkpoints.
Checkout [the docs](https://nightlies.apache.org/flink/flink-docs-release-1.14/docs/deployment/memory/network_mem_tuning/#the-buffer-debloating-mechanism)

> to see how to activate this feature.

Reviewer: ```suggestion
## Buffer Debloating

*Buffer Debloating* is a new feature in Flink that automatically tunes
network memory usage to ensure high throughput while minimizing in-flight data. The goal is to minimize
checkpoint latency and cost.

Flink buffers a certain amount of data in its network stack in order to leverage the
bandwidth of fast networks. A Flink application running with high throughput uses some of
that memory. Aligned checkpoints flow with the data through the network buffers in milliseconds.

When a Flink application becomes (temporarily) backpressured (i.e. when being backpressured
by an external system, or when hitting skewed records), this typically results in a lot more data in
the network buffers than is necessary for the network to support the application's current throughput
(which is lowered due to backpressure). There is even an adverse effect: more buffered data means that
the checkpoints need to do more work. Aligned checkpoint barriers need to wait for more data to be
processed while unaligned checkpoints need to persist more in-flight data.

When there is more data in the network buffers than necessary because of backpressure,
*Buffer Debloating* comes into play. It changes the network stack from keeping up to X bytes of data
to keeping data that is worth X milliseconds of receiver computing time. With a default setting
of 1000 milliseconds, that means the network stack will buffer as much data as the receiving task can
process in 1000 milliseconds. These values are constantly measured and adjusted, so that the system can maintain these properties even under varying conditions. Under backpressure, Flink can now provide
stable and predictable alignment times for aligned checkpoints and can vastly
reduce the amount of in-flight data stored in unaligned checkpoints.

<figure style="align-content: center">
  <img src="{{ site.baseurl }}/img/blog/2021-09-25-release-1.14.0/buffer_debloating.png" style="display: block; margin-left: auto; margin-right: auto; width: 600px"/>
</figure>

Buffer Debloating acts as a complementary feature, or even alternative, to unaligned checkpoints.
Check out [the documentaion](https://nightlies.apache.org/flink/flink-docs-release-1.14/docs/deployment/memory/network_mem_tuning/#the-buffer-debloating-mechanism)
to learn how to activate this feature.
```

## Fine-grained Resource Management

> The new *Fine-grained Resource Management* is an advanced feature to increase the resource
> utilization of large shared clusters.
> 
> Flink clusters execute various data processing workloads, and different data processing steps typically need
> different resources: compute resource, memory, etc. For example, most `map()` functions are fairly
> lightweight, but large windows with long retention can benefit from lot's or memory.
> By default, Flink manages resources in coarse-grained units called *slots*, which are slices
> of a TaskManager's resources. By default, Streaming pipelines fill a slot with one parallel
> subtask of each operator, so each slot holds a pipeline of subtasks.
> Through *'slot sharing groups'*, users can influence how subtasks are assigned to slots.
> 
> With fine-grained resource management, TaskManager slots can now be dynamically sized.
> Transformations and Operators specify what resource profiles they would like (size of CPU,
> memory pools, disk space) and Flink's Resource Manager and TaskManagers slice off that specific
> part of a TaskManagers total resources. Think of it as a minimal lightweight resource orchestration
> layer within Flink. The figure below illustrates the difference between the existing default way (shared
> fixed-size slots) and the new fine-grained resource management.

Reviewer: ```suggestion
*Fine-grained resource management* is an advanced new feature that increases the resource
utilization of large shared clusters.

Flink clusters execute various data processing workloads. Different data processing steps typically need
different resources such as compute resources and memory. For example, most `map()` functions are fairly
lightweight, but large windows with long retention can benefit from lots of memory.
By default, Flink manages resources in coarse-grained units called *slots*, which are slices
of a TaskManager's resources. Streaming pipelines fill a slot with one parallel
subtask of each operator, so each slot holds a pipeline of subtasks.
Through *'slot sharing groups'*, users can influence how subtasks are assigned to slots.

With fine-grained resource management, TaskManager slots can now be dynamically sized.
Transformations and operators specify what resource profiles they would like (CPU size,
memory pools, disk space) and Flink's Resource Manager and TaskManagers slice off that specific
part of a TaskManager's total resources. You can think of it as a minimal lightweight resource orchestration
layer within Flink. The figure below illustrates the difference between the current default mode of shared
fixed-size slots and the new fine-grained resource management feature.
```

<figure style="align-content: center">
  <img src="{{ site.baseurl }}/img/blog/2021-09-25-release-1.14.0/fine_grained_resource_management.png" style="display: block; margin-left: auto; margin-right: auto; width: 600px"/>
</figure>

> Hmmmm, why would we even do that, is this not better handled directly by Kubernetes or Yarn?
> For many cases, that is in fact true. There are, however, some reasons to not rely purely on K8s/Yarn for this:
> 
>   - For many small slots, the overhead of dedicated TMs is very high (JVM overhead, Flink control data structures).
>     Slot-sharing works around this implicitly by sharing the slots between all operator types, which means
>     sharing resources between lightweight (need small slots) and heavyweight (need large slots) operators.
>     However, this works only well when all operators share the same parallelism, which isn't aways optiomal.
>     Furthermore, sometimes certain operators work better when run in isolation (for example ML training operators
>     that need dedicated GPU resources).
>   - K8s and Yarn take often quite some time to fulfill requests, especially on loaded clusters.
>     For certain batch jobs, a of efficiency gets lost waiting for Yarn and K8s to fullfill the requests.
>     
> Who should use this feature? For most jobs (both streaming and batch) the default resource management mechanism
> are perfectly fine. This feature can help you increase the resource efficiency, if you have either long-running
> streaming jobs, or fast batch jobs, where different stages have different resource requirements, and you possibly
> already tuned the parallelism of different operators to different values. Then this feature can help you
> increase the resource efficiency.
> Alibaba's internal Flink-based platform has used this mechanism for some time now and has significantly
> increased the resource utilization of the cluster. 
> 
> Please refer to the [Fine-grained Resource Management Docs](https://nightlies.apache.org/flink/flink-docs-release-1.14/docs/deployment/finegrained_resource/)
> for details on how to use this feature.

Reviewer: ```suggestion
You may be wondering whether this would be better handled directly by Kubernetes or YARN. 
Here are some reasons why you may not want to rely purely on these popular resource managers:

  - For many small slots, the overhead of dedicated TaskManagers is very high (JVM overhead, Flink control data
    structures). Slot-sharing implicitly works around this by sharing the slots between all operator types, which    
    means sharing resources between lightweight operators (which need small slots) and heavyweight operators   
    (which need large slots). However, this only works well when all operators share the same parallelism, which is 
    not always optimal. Furthermore, certain operators sometimes work better when run in isolation (i.e. training  
    operators for machine learning that need dedicated GPU resources).
  - Kubernetes and YARN often take quite some time to fulfill requests, especially on loaded clusters.
    For certain batch jobs, efficiency gets lost while waiting for the requests to be fulfilled. 
    
So when should you use this feature? For most batch and streaming jobs, the default resource management mechanism is perfectly suitable. Fine-grained resourced management can help you increase resource efficiency if you have either long-running streaming jobs or fast batch jobs, where different stages have different resource requirements, and you may have already tuned the parallelism of different operators to different values. 
Alibaba's internal Flink-based platform has used this mechanism for some time now and the resource utilization of the cluster has significantly increased. 

Please refer to the [Fine-grained Resource Management documentation](https://nightlies.apache.org/flink/flink-docs-release-1.14/docs/deployment/finegrained_resource/)
for details on how to use this feature.
```

# Connectors

## Connector metrics

Flink 1.14 standardized the metrics for connectors ([FLIP-33](https://cwiki.apache.org/confluence/display/FLINK/FLIP-33%3A+Standardize+Connector+Metrics))
The community will gradually pull them through all connectors, as we rework them
onto the new unified APIs over the next releases. In Flink 1.14, we cover the Kafka connector,
and partially the File connectors.

Connectors are the entry and the exit point for data in Flink job. If a job is not running as
expected, the connector telemetry is among first parts to check. We believe this will become

> a nice improvement for everyone operating production Flink applications.

Reviewer: ```suggestion
## Connector Metrics

Metrics for connectors have been standardized in this release (see [FLIP-33](https://cwiki.apache.org/confluence/display/FLINK/FLIP-33%3A+Standardize+Connector+Metrics)).
The community will gradually pull metrics through all connectors, as we rework them
onto the new unified APIs over the next releases. In Flink 1.14, we cover the Kafka connector
and (partially) the FileSystem connectors.

Connectors are the entry and exit points for data in a Flink job. If a job is not running as
expected, the connector telemetry is among the first parts to be checked. We believe this will become
a nice improvement when operating Flink applications in production.
```

## Pulsar Source

> Flink added a [new Source and Sink](https://nightlies.apache.org/flink/flink-docs-release-1.14/docs/connectors/datastream/pulsar/).

Reviewer 4: It seems like Pulsar connector don't support Sink API in 1.14 release.

> Flink added a [new Source and Sink](https://nightlies.apache.org/flink/flink-docs-release-1.14/docs/connectors/datastream/pulsar/).
> for [Apache Pulsar](https://pulsar.apache.org/). Pulsar is a distributed messaging and streaming
> system.
> 
> 
> 
> In this release, Flink added the [Apache Pulsar](https://pulsar.apache.org/) connector.
> The Pulsar connector reads data from Pulsar topics and supports both streaming and batch execution modes.
> With the support of the transaction functionality (introduced in Pulsar 2.8.0), the Pulsar connector provides
> exactly-once delivery semantic to ensure that a message is delivered exactly once to the consumer,
> even if a producer retries sending the message.
> 
> To support the different message-ordering and scaling requirements of different use cases, the Pulsar
> source connector exposes four subscription types:
>   - [Exclusive](https://pulsar.apache.org/docs/en/concepts-messaging/#exclusive)
>   - [Shared](https://pulsar.apache.org/docs/en/concepts-messaging/#shared)
>   - [Failover](https://pulsar.apache.org/docs/en/concepts-messaging/#failover)
>   - [Key-Shared](https://pulsar.apache.org/docs/en/concepts-messaging/#key_shared)
> 
> The connector currently supports the DataStream API. Table API/SQL bindings are expected to be
> contributed in a future release. For details about how to use the Pulsar connector, see
> [Apache Pulsar Connector](https://nightlies.apache.org/flink/flink-docs-release-1.14/docs/connectors/datastream/pulsar/#apache-pulsar-connector).

Reviewer: ```suggestion
In this release, we added the [Apache Pulsar](https://pulsar.apache.org/) source connector.
The Pulsar connector reads data from Pulsar topics and supports both streaming and batch execution modes.
With the support of the transaction functionality (introduced in Pulsar 2.8.0), the Pulsar connector provides
exactly-once delivery semantic to ensure that a message is delivered exactly once to the consumer,
even if a producer retries sending the message.

To support the different message-ordering and scaling requirements of different use cases, the Pulsar
source connector exposes four subscription types:
  - [Exclusive](https://pulsar.apache.org/docs/en/concepts-messaging/#exclusive)
  - [Shared](https://pulsar.apache.org/docs/en/concepts-messaging/#shared)
  - [Failover](https://pulsar.apache.org/docs/en/concepts-messaging/#failover)
  - [Key-Shared](https://pulsar.apache.org/docs/en/concepts-messaging/#key_shared)

The connector currently supports the DataStream API. Table API/SQL bindings are expected to be
contributed in a future release. For details about how to use the Pulsar connector, see
[Apache Pulsar Connector](https://nightlies.apache.org/flink/flink-docs-release-1.14/docs/connectors/datastream/pulsar/#apache-pulsar-connector).
```

# PyFlink

> ## Performance through Chaining
> 
> Flink's Python DataStream API now chains Python functions, which provides a significant
> boost to PyFlink's overall performance. 
> 
> Similar to how the Java APIs chain transformation functions/operators within a task to avoid
> serialization overhead, PyFlink now chains Python functions. In PyFlink's case, the
> chaining not only eliminates serialization overhead, but also reduces RPC round trips
> between the Java and Python processes.
> Python function chaining was already available for Python UDFs used in the Table API & SQL.
> In Flink 1.14, chaining is also exploited for the cPython functions in Python DataStream API.

Reviewer: ```suggestion
## Performance Improvement through Chaining

Similar to how the Java APIs chain transformation functions/operators within a task to avoid
serialization overhead, PyFlink now chains Python functions. In PyFlink's case, the
chaining not only eliminates serialization overhead, but also reduces RPC round trips
between the Java and Python processes. This provides a significant
boost to PyFlink's overall performance. 

Python function chaining was already available for Python UDFs used in the Table API & SQL.
In Flink 1.14, chaining is also exploited for the cPython functions in Python DataStream API.
```

## Loopback Mode for Debugging

Python functions are normally executed in a separate Python process next to Flink's JVM.
This architecture makes it difficult to debug Python code.

> PyFlink 1.14 introduces a *loopback mode*, which is activated by default for local deployments.
> In that mode, user-defined Python functions will be executed in the Python process of the client,
> which is the entry point process that starts the PyFlink program and contains the DataStream and
> Table API code that builds the dataflow DAG. Users can debug their Python functions now easily
> by simply setting breakpoints in the IDEs when launching a PyFlink job locally.

Reviewer: ```suggestion
PyFlink 1.14 introduces a *loopback mode*, which is activated by default for local deployments.
In this mode, user-defined Python functions will be executed in the Python process of the client,
which is the entry point process that starts the PyFlink program and contains the DataStream API and
Table API code that builds the dataflow DAG. Users can now easily debug their Python functions
by setting breakpoints in their IDEs when launching a PyFlink job locally.
```

## Miscellaneous Improvements

> There are also many other improvements to PyFlink, such as support for executing
> jobs in YARN application mode, or support for compressed tgz files as Python archives.
> Check out the [Python API docs](https://nightlies.apache.org/flink/flink-docs-release-1.14/docs/dev/python/overview/) 
> for details.

Reviewer: ```suggestion
There are also many other improvements to PyFlink, such as support for executing
jobs in YARN application mode and support for compressed tgz files as Python archives.
Check out the [Python API documentation](https://nightlies.apache.org/flink/flink-docs-release-1.14/docs/dev/python/overview/) 
for more details.
```

> # Goodbye Legacy SQL Engine and Mesos
> 
> Maintaining an open source project also means saying good-bye to some beloved features eventually.
> 
> When we added the Blink SQL Engine to Flink more than two years ago, it was clear that it would
> eventually replace the previous SQL engine. Blink was faster and more feature complete.
> For a year now, Blink has been the default SQL engine. With Flink 1.14 we finally remove all
> code from the previous SQL engine. Not only does this help contributors to be faster with their
> development, but it means we can also drop a plethora of outdated interfaces and reduce confusion
> for users which interfaces to use when implementing custom connectors or functions.
> 
> The active integration with Apache Mesos was also removed, because we saw very little interest by
> users and no contributors interested in maintaining this part of the system. Flink 1.14 can no
> longer run on Mesos without the help of projects like Marathon, and the Flink Resource Manager
> can no longer request and release resources from Mesos for workloads with changing changing
> resource requirements.

Reviewer: ```suggestion
# Goodbye to the Legacy SQL Engine and to Mesos Support

Maintaining an open source project means sometimes saying goodbye to some beloved features.

When we added the Blink SQL Engine to Flink more than two years ago, it was clear that it would
eventually replace the previous SQL engine. Blink was faster and more feature-complete.
For a year now, Blink has been the default SQL engine. With Flink 1.14, we finally removed all legacy
code from the previous SQL engine. This helps contributors develop faster. It allowed us to drop many outdated interfaces and reduce confusion for users about which interfaces to use when implementing custom connectors or functions.

The active integration with Apache Mesos was also removed because we saw little interest by
users in this feature and few contributors were interested in maintaining this part of the system. Flink 1.14 can no
longer run on Mesos without the help of projects like Marathon, and the Flink Resource Manager
can no longer request and release resources from Mesos for workloads with changing
resource requirements.
```

# Upgrade Notes

While we aim to make upgrades as smooth as possible, some of the changes require users
to adjust some parts of the program when upgrading to Apache Flink 1.14.
Please take a look at the [release notes](https://nightlies.apache.org/flink/flink-docs-release-1.14/release-notes/flink-1.14/)

> for a list of adjustemnts and issues to check during upgrades.

Reviewer: ```suggestion
for a list of adjustments to make and issues to check during upgrades.
```
